python_modules/dagster/dagster/_core/definitions/partition_mapping.py

```diff
@@ -320,6 +328,124 @@ def get_downstream_partitions_for_partitions(
+        downstream_partition_keys_by_upstream_partition_key (Dict[str, str | Collection[str]]):
+          The single or multi-valued correspondence from upstream keys to downstream keys.
+    """
+
+    def __init__(
+        self,
+        downstream_partition_keys_by_upstream_partition_key: Mapping[
+            str, Union[str, Collection[str]]
+        ],
+    ):
+        check.mapping_param(
+            downstream_partition_keys_by_upstream_partition_key,
+            "downstream_partition_keys_by_upstream_partition_key",
+            key_type=str,
+            value_type=(str, collections.abc.Collection),
+        )
+
+        # cache forward and reverse mappings
+        self._mapping = defaultdict(set)
+        for (
+            upstream_key,
+            downstream_keys,
+        ) in downstream_partition_keys_by_upstream_partition_key.items():
+            self._mapping[upstream_key] = (
+                {downstream_keys} if isinstance(downstream_keys, str) else set(downstream_keys)
+            )
+
+        self._inverse_mapping = defaultdict(set)
+        for upstream_key, downstream_keys in self._mapping.items():
+            for downstream_key in downstream_keys:
+                self._inverse_mapping[downstream_key].add(upstream_key)
+
+    @cached_method
+    def _check_upstream(self, *, upstream_partitions_def: StaticPartitionsDefinition):
+        """
+        validate that the mapping from upstream to downstream is only defined on upstream keys
+        """
+        check.inst(
+            upstream_partitions_def,
+            StaticPartitionsDefinition,
+            "StaticPartitionMapping can only be defined between two StaticPartitionsDefinitions",
+        )
+        upstream_keys = upstream_partitions_def.get_partition_keys()
+        extra_keys = set(self._mapping.keys()).difference(upstream_keys)
+        if extra_keys:
+            raise ValueError(
+                f"mapping source partitions not in the upstream partitions definition: {extra_keys}"
+            )
+
+    @cached_method
+    def _check_downstream(self, *, downstream_partitions_def: StaticPartitionsDefinition):
+        """
+        validate that the mapping from upstream to downstream only maps to downstream keys
+        """
+        check.inst(
+            downstream_partitions_def,
+            StaticPartitionsDefinition,
+            "StaticPartitionMapping can only be defined between two StaticPartitionsDefinitions",
+        )
+        downstream_keys = downstream_partitions_def.get_partition_keys()
+        extra_keys = set(self._inverse_mapping.keys()).difference(downstream_keys)
+        if extra_keys:
+            raise ValueError(
+                f"mapping target partitions not in the downstream partitions definition: {extra_keys}"
+            )
+
+    def get_downstream_partitions_for_partitions(
+        self,
+        upstream_partitions_subset: PartitionsSubset,
+        downstream_partitions_def: StaticPartitionsDefinition,
```

Comment:
To satisfy mypy, I think you need to change this to 
```python
        downstream_partitions_def: StaticPartitionsDefinition,
```

Here's the complaint:
```
[2023-01-07T01:51:57Z] dagster/_core/definitions/partition_mapping.py:415: error: Argument 2 of "get_upstream_partitions_for_partitions" is incompatible with supertype "PartitionMapping"; supertype defines the argument type as "PartitionsDefinition[Any]"  [override]
--
  | [2023-01-07T01:51:57Z] dagster/_core/definitions/partition_mapping.py:415: note: This violates the Liskov substitution principle
  | [2023-01-07T01:51:57Z] dagster/_core/definitions/partition_mapping.py:415: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

Same on line 418.

Reply:
Done.

I'm not clear on dagster's preferred etiquite for resolving conversations so I've left them open.

@clairelin135 I'm not sure how I removed you as a reviewer. Apologies, and I cannot see how to add you back.